hooks/post-command

```diff
@@ -33,90 +33,145 @@ display_success() {
+
+        local status=$?
+        if [[ $status -ne 0 ]]; then
+            display_error "docker inspect" "Failed to get the local image digest, will continue using supplied image reference ${image}"
+        else
+            display_success "docker inspect" "Will continue using ${digest}"
+            image="${digest}"
+        fi
+    fi
+}
 
-# Parameters
-############
+# Common parameters
+###################
 
-# This is a required parameter
+# image is a required parameter
 image=${BUILDKITE_PLUGIN_COSIGN_IMAGE}
 if [[ -z "${image}" ]]; then
-    fail_with_message "cosign" "No image specified"
+    fail_with_message "cosign" "Image not specified"
 fi
+use_image_digest
+
+# flags for the cosign sign command
+sign_flags=("-y" "--output-signature" "out.sig")
 
 is_keyless=${BUILDKITE_PLUGIN_COSIGN_KEYLESS:-true}
 
 # Hook functions
 ################
 
-cosign_keyless() {
-    local fulcio_url=${BUILDKITE_PLUGIN_COSIGN_KEYLESS_CONFIG_FULCIO_URL:-"https://fulcio.sigstore.dev"}
-    local rekor_url=${BUILDKITE_PLUGIN_COSIG_KEYLESS_CONFIGN_REKOR_URL:-"https://rekor.sigstore.dev"}
-    local oidc_issuer=${BUILDKITE_PLUGIN_COSIG_KEYLESS_CONFIGN_OIDC_ISSUER:-"https://oauth2.sigstore.dev/auth"}
-    local oidc_provider=${BUILDKITE_PLUGIN_COSIG_KEYLESS_CONFIGN_OIDC_PROVIDER:-"buildkite-agent"}
-    
-    echo "--- :key: Cosign keyless signing"
-
-    rm out.sig || true
-
-    COSIGN_EXPERIMENTAL=1 cosign sign \
-        -y \
-        --fulcio-url="${fulcio_url}" \
-        --rekor-url="${rekor_url}" \
-        --oidc-issuer="${oidc_issuer}" \
-        --oidc-provider="${oidc_provider}" \
-        --output-signature=out.sig \
-        "${image}"
+# if provided, initialise cosign with a custom TUF configuration
+cosign_init() {
+    echo "--- :key: Init cosign"
 
-    status=$?
-    if [[ $status -ne 0 ]]; then
-        fail_with_message "cosign" "Failed to sign image"
+    # flags for the cosign initialize command
+    local init_flags=()
+
+    if [[ "${is_keyless}" == true ]]; then
+        local tuf_mirror_url=${BUILDKITE_PLUGIN_COSIGN_KEYLESS_CONFIG_TUF_MIRROR_URL}
+        local tuf_root_url=${BUILDKITE_PLUGIN_COSIGN_KEYLESS_CONFIG_TUF_ROOT_URL}
+    else
+        local tuf_mirror_url=${BUILDKITE_PLUGIN_COSIGN_KEYED_CONFIG_TUF_MIRROR_URL}
+        local tuf_root_url=${BUILDKITE_PLUGIN_COSIGN_KEYED_CONFIG_TUF_ROOT_URL}
     fi
 
-    local signature=$(cat out.sig)
+    if [[ -n "${tuf_mirror_url}" ]]; then
+        init_flags+=("--mirror" "${tuf_mirror_url}")
+    fi
 
-    display_success "cosign" "Successfully signed image."
-    cat <<EOF | buildkite-agent annotate --style success --context "cosign-signature"
-### Signed image
-\`\`\`
-$image
-\`\`\`
+    if [[ -n "${tuf_root_url}" ]]; then
+        init_flags+=("--root" "${tuf_root_url}")
+    fi
 
-### Signature
-\`\`\`
-$signature
-\`\`\`
-EOF
+    if [ ${#init_flags[@]} -gt 0 ]; then
+        rm -rf ~/.sigstore
```

Comment:
Does this mean this plugin can only run serially on a given host? Is there a variable one can set to control this directory? Would be ideal if it were unique per plugin run.

Reply:
good point, thanks!

looks like that's not directly exposed/customisable by `cosign initialize` (via [flags](https://github.com/sigstore/cosign/blob/2b538f8963b44da87e8b9f18b44ecdbfa6c238ce/cmd/cosign/cli/options/initialize.go#L31-L39)) but the sigstore `tuf client` reads the `TUF_ROOT` [env var](https://github.com/sigstore/sigstore/blob/b777e4be352ebf9394d534271f3dd888908e839a/pkg/tuf/client.go#L53-L54) to figure out the location, so i'll try to use that:

> // TufRootEnv is the name of the environment variable that locates an alternate local TUF root location.
TufRootEnv = "TUF_ROOT"

=> https://github.com/sigstore/sigstore/blob/b777e4be352ebf9394d534271f3dd888908e839a/pkg/tuf/client.go#L559

similarly, i've changed made `out.sig` "dynamic" (pseudo-random) for the same reasons

Reply:
yep, looks like that worked:

```
🔑 Init cosign
Root status:
 {
	"local": "/var/lib/buildkite-agent/.sigstore-13508/root",
	"remote": "https://<REDACTED>",
	"metadata": {
		"root.json": {
			"version": 1,
			"len": 2178,
			"expiration": "01 Dec 24 19:08 UTC",
			"error": ""
		},
		"snapshot.json": {
			"version": 1,
			"len": 618,
			"expiration": "01 Dec 24 19:08 UTC",
			"error": ""
		},
		"targets.json": {
			"version": 1,
			"len": 1373,
			"expiration": "01 Dec 24 19:08 UTC",
			"error": ""
		},
		"timestamp.json": {
			"version": 1,
			"len": 619,
			"expiration": "01 Dec 24 19:08 UTC",
			"error": ""
		}
	},
	"targets": [
		"ctfe.pub",
		"fulcio_v1.crt.pem",
		"rekor.pub"
	]
}
Successfully initialised
```

Reply:
perhaps we could pr to update `cosign initialize` cmd's [docs](https://github.com/sigstore/cosign/blob/2b538f8963b44da87e8b9f18b44ecdbfa6c238ce/doc/cosign_initialize.md?plain=1#L16) and [code](https://github.com/sigstore/cosign/blob/2b538f8963b44da87e8b9f18b44ecdbfa6c238ce/cmd/cosign/cli/initialize.go#L40) so users are aware they can set tuf's root directory

Reply:
fwiw, I used @prezha's `TUF_ROOT` environment variable finding in a separate project to run multiple instances of `cosign` simultaneously on the same computer. Prior to making that change, I was experiencing "resource busy" errors due to each cosign process modifying / writing files in `~/.sigstore`. I think this emulates the same kind of situation we would see in a Buildkite CI runner.

In other words: @prezha's change appears to prevent cosign from stepping on other cosign processes.

Reply:
thanks @sfox-equinix for checking and confirming!

@JAORMX is there anything additional you think should be addressed?